lib/sql.ts

```diff
@@ -0,0 +1,147 @@
+			statement = statements.get(num);
+			results = statement?.get(data);
+		}
+			break;
+		case 'run': {
+			const {num, data} = message;
+			statement = statements.get(num);
+			results = statement?.run(data);
+		}
+			break;
+		case 'exec': {
+			const {data} = message;
+			database.exec(data);
+		}
+			break;
+		}
+		parentPort!.postMessage(results);
+	});
+}
+
+export class SQLDatabase extends Worker {
+	statements: Map<string, number>;
+	constructor(options: SQLOptions) {
+		super(__filename, {workerData: options});
+		this.statements = new Map();
+	}
+	async prepare(statement: string) {
+		const cachedStatement = this.statements.get(statement);
+		if (cachedStatement) {
+			return cachedStatement;
+		}
+		const int = await this.makeRequest({type: 'prepare', data: statement});
+		if (int) this.statements.set(statement, int);
+		return int;
+	}
+	async all(statement: string | number, data: InputType = {}) {
+		const num = typeof statement === 'number' ? statement : this.statements.get(statement);
+		if (num === undefined || ![...this.statements.values()].includes(num)) {
+			throw new Error(`Prepare a statement before using another database function with SQLDatabase.prepare.`);
+		}
+		const result = await this.makeRequest({type: 'all', num: num, data});
+		return result;
+	}
+	async get(statement: string | number, data: InputType = {}) {
+		const num = typeof statement === 'number' ? statement : this.statements.get(statement);
+		if (num === undefined || ![...this.statements.values()].includes(num)) {
+			throw new Error(`Prepare a statement before using another database function with SQLDatabase.prepare.`);
+		}
+		const result = await this.makeRequest({type: 'get', data, num});
+		return result;
+	}
+	async run(statement: string | number, data: InputType) {
+		const num = typeof statement === 'number' ? statement : this.statements.get(statement);
+		if (num === undefined || ![...this.statements.values()].includes(num)) {
+			throw new Error(`Prepare a statement before using another database function with SQLDatabase.prepare.`);
+		}
+		const result = await this.makeRequest({type: 'run', num, data});
+		return result;
+	}
+	/** We don't want these to be prepared statements, since this should be used SPARINGLY. */
+	async exec(statement: string) {
+		const result = await this.makeRequest({type: 'exec', data: statement});
+		return result;
+	}
+	private async makeRequest(args: Message) {
+		this.postMessage(args);
+		const resultPromise = new Promise<any>((resolve, reject) => {
+			this.once('message', result => {
+				resolve(result);
+			});
+			this.once('error', err => {
+				reject(err);
+			});
+		});
+		const result = await resultPromise;
+		return result;
+	}
+}
+/**
+ * @param options Either an object of filename, extension, or just the string filename
+ */
+export function SQL(options: SQLOptions | string) {
+	if (typeof options === 'string') options = {file: options};
+	return new SQLDatabase(options);
```

Comment:
A new worker for each database...? Wouldn't something like a thread pool be more optimal here?